components/suggest/src/store.rs

```diff
@@ -0,0 +1,421 @@
+/// user types into the address bar, and periodically call `ingest` in the
+/// background to update the database with new suggestions from Remote Settings.
+///
+/// The store keeps track of the state needed to support incremental ingestion,
+/// but doesn't schedule the ingestion work itself, or decide how many
+/// suggestions to ingest at once. This is for two reasons:
+///
+/// 1. The primitives for scheduling background work vary between platforms. You
+///    might use an idle timer on Desktop, `WorkManager` on Android, or
+///    `BGTaskScheduler` on iOS.
+/// 2. Ingestion limits can change, depending on the platform and the needs of
+///    the application. A mobile device on a metered connection might want to
+///    request a small subset of the Suggest data and fetch the rest later,
+///    while a desktop on a fast link might request the entire dataset on first
+///    launch.
+pub struct SuggestStore {
+    path: PathBuf,
+    dbs: OnceCell<SuggestStoreDbs>,
+    settings_client: remote_settings::Client,
+}
+
+/// Limits for an ingestion from Remote Settings.
+pub struct IngestLimits {
+    /// The maximum number of records to request from Remote Settings.
+    /// Each record has about 200 suggestions.
+    pub records: Option<u64>,
+}
+
+impl SuggestStore {
+    /// Creates a Suggest store.
+    pub fn new(
+        path: &str,
+        settings_config: Option<RemoteSettingsConfig>,
+    ) -> SuggestApiResult<Self> {
+        Ok(Self::new_inner(path, settings_config)?)
+    }
+
+    fn new_inner(
+        path: impl AsRef<Path>,
+        settings_config: Option<RemoteSettingsConfig>,
+    ) -> Result<Self> {
+        let settings_client = remote_settings::Client::new(settings_config.unwrap_or_else(|| {
+            RemoteSettingsConfig {
+                server_url: None,
+                bucket_name: None,
+                collection_name: RS_COLLECTION.into(),
+            }
+        }))?;
+        Ok(Self {
+            path: path.as_ref().into(),
+            dbs: OnceCell::new(),
+            settings_client,
+        })
+    }
+
+    /// Returns this provider's database connections, initializing them if
+    /// they're not already open.
+    fn dbs(&self) -> Result<&SuggestStoreDbs> {
+        self.dbs
+            .get_or_try_init(|| SuggestStoreDbs::open(&self.path))
+    }
+
+    /// Queries the database for suggestions that match the `keyword`.
+    pub fn query(&self, query: &SuggestionQuery) -> SuggestApiResult<Vec<Suggestion>> {
+        if query.keyword.is_empty() {
+            return Ok(Vec::new());
+        }
+        Ok(self
+            .dbs()?
+            .reader
+            .fetch_by_keyword(&query.keyword)?
+            .into_iter()
+            .filter(|suggestion| {
+                (suggestion.is_sponsored && query.include_sponsored)
+                    || (!suggestion.is_sponsored && query.include_non_sponsored)
+            })
```

Comment:
For sure!

The short answer: I didn't want to complicate `fetch_by_keyword` too much at this stage by doing the equivalent of the `!NONSPONSORED_IAB_CATEGORIES.contains(&iab_category)` check in SQL.

The longer answer: We'll almost certainly be making a big schema change that'll affect `fetch_by_keyword` and `is_sponsored`, and it's not clear that we'll actually want to filter suggestions this way.

Right now, we're storing everything in one big `suggestions` table, but I think we'll want to partition the suggestions into separate tables for each "provider": adM, Wikipedia, and soon MDN, AMO, and Pocket. adM and Wikipedia suggestions currently use the same format, even though a lot of the adM fields aren't relevant for Wikipedia, and MDN / AMO / Pocket suggestions will use a different format entirely.

Since we'll be changing the schema and the query, anyway, I went with the quick approach here 😅

Reply:
This makes total sense - thanks!